flink-runtime/src/main/java/org/apache/flink/runtime/util/SetArrayDeque.java

```diff
@@ -16,20 +16,38 @@
  * limitations under the License.
  */
 
-package org.apache.flink.runtime.io.network.partition.consumer;
+package org.apache.flink.runtime.util;
+
+import java.util.ArrayDeque;
+import java.util.HashSet;
+import java.util.Set;
 
 /**
- * Listener interface implemented by consumers of {@link InputGate} instances
- * that want to be notified of availability of buffer or event instances.
+ * Wrapper of {@link ArrayDeque} with quick {@link HashSet}'s based {@link #contains(Object)} method.
  */
-public interface InputGateListener {
+public class SetArrayDeque<T> {
+	private final ArrayDeque<T> deque = new ArrayDeque<>();
+	private final Set<T> set = new HashSet<>();
+
+	public boolean add(T element) {
+		return deque.add(element) || set.add(element);
```

Comment:
I can change it to `void add(T)` ;) But ok, I will add the unit test for this.

Reply:
It is not only about the return value, but even worse you can add duplicates to the dequeue because it is ignored that the value is potentially already in the set. So this is not set semantics anymore.

Reply:
I meant that if I change the signature to `void add(T)` I can drop the `||` expression and solve the problem.

Btw, in this case `&&` would also work fine, since `deque.add()` always returns true.